src/primitives/triangle/scanline_intersections.rs

```diff
@@ -0,0 +1,230 @@
+    /// Represents part of the stroke.
+    Stroke,
+
+    /// Represents the interior of the shape.
+    Fill,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+struct LineConfig {
+    first: Option<Line>,
+    second: Option<Line>,
+    internal: Option<(Line, PointType)>,
+}
+
+/// Triangle scanline intersections iterator.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub(in crate::primitives::triangle) struct ScanlineIntersections {
+    lines: LineConfig,
+    triangle: Triangle,
+    stroke_width: u32,
+    stroke_offset: StrokeOffset,
+    has_fill: bool,
+}
+
+static EMPTY: Triangle = Triangle::new(Point::zero(), Point::zero(), Point::zero());
+
+impl ScanlineIntersections {
+    /// Create a new thick segments iterator.
+    pub fn new(
+        triangle: &Triangle,
+        stroke_width: u32,
+        stroke_offset: StrokeOffset,
+        has_fill: bool,
+        scanline_y: i32,
+    ) -> Self {
+        if let Some(lines) =
+            generate_lines(triangle, stroke_width, stroke_offset, has_fill, scanline_y)
+        {
+            Self {
+                lines,
+                has_fill,
+                triangle: *triangle,
+                stroke_width,
+                stroke_offset,
+            }
+        } else {
+            Self::empty()
+        }
+    }
+
+    /// Empty.
+    pub fn empty() -> Self {
+        Self {
+            lines: LineConfig::default(),
+            has_fill: false,
+            triangle: EMPTY,
+            stroke_width: 0,
+            stroke_offset: StrokeOffset::None,
+        }
+    }
+
+    /// Reset with a new scanline.
+    pub fn reset_with_new_scanline(&mut self, scanline_y: i32) {
+        if let Some(lines) = generate_lines(
+            &self.triangle,
+            self.stroke_width,
+            self.stroke_offset,
+            self.has_fill,
+            scanline_y,
+        ) {
+            self.lines = lines
+        }
+    }
+}
+
+fn generate_lines(
+    triangle: &Triangle,
+    stroke_width: u32,
+    stroke_offset: StrokeOffset,
+    has_fill: bool,
+    scanline_y: i32,
+) -> Option<LineConfig> {
+    let mut edge_intersections = {
+        let mut idx = 0;
+        let mut left = None;
+        let mut right = None;
+
+        let t = triangle;
+
+        core::iter::from_fn(move || {
+            if stroke_width == 0 {
+                return None;
+            }
+
+            while idx < 3 {
+                let start = LineJoin::from_points(
+                    t.vertex(idx),
+                    t.vertex(idx + 1),
+                    t.vertex(idx + 2),
+                    stroke_width,
+                    stroke_offset,
+                );
+                let end = LineJoin::from_points(
+                    t.vertex(idx + 1),
+                    t.vertex(idx + 2),
+                    t.vertex(idx + 3),
+                    stroke_width,
+                    stroke_offset,
+                );
```

Comment:
Something I'd like to try to improve performance with in the future is to cache the 3 generated edge `ThickSegment`s. I chose not to do it in this PR because it adds `72 * 3 = 216` bytes to the iterator size (minus a few bytes when removing the `LineConfig` struct) which is pretty unacceptable, even for a short-lived iterator struct IMO.

Reply:
Sounds interesting, do you want to open an issue for this?

Reply:
I don't think so. I want to take a closer look at all primitive drawing performance at some point, so I'll fold it into that.

If I do try this trick in the future, do you think 216+ bytes for the iterator alone is too much?

Reply:
> If I do try this trick in the future, do you think 216+ bytes for the iterator alone is too much?

It's a lot and we should IMO first see if we can gain performance in other ways. If this is the only way to get a substantial performance gain we might need to consider to add a feature to select between better performance or lower memory consumption. I'm sure there would be more instances where such a feature could be useful.

Reply:
Yeah, I'll try other ways first.

> we might need to consider to add a feature to select between better performance or lower memory consumption. 

I'm glad we're thinking of the same things :)